app/assets/javascripts/discourse/lib/site-settings-plugin.js

```diff
@@ -0,0 +1,85 @@
+const Plugin = require("broccoli-plugin");
+const Yaml = require("js-yaml");
+const fs = require("fs");
+const concat = require("broccoli-concat");
+const mergeTrees = require("broccoli-merge-trees");
+const deepmerge = require("deepmerge");
+const glob = require("glob");
+const { shouldLoadPluginTestJs } = require("discourse/lib/plugin-js");
+
+let built = false;
+
+class SiteSettingsPlugin extends Plugin {
+  constructor(inputNodes, inputFile, options) {
+    super(inputNodes, {
+      ...options,
+      persistentOutput: true,
+    });
+  }
+
+  build() {
+    if (built) {
+      return;
+    }
+
+    let parsed = {};
+
+    this.inputPaths.forEach((path) => {
+      let inputFile;
+      if (path.includes("plugins")) {
+        inputFile = "settings.yml";
+      } else {
+        inputFile = "site_settings.yml";
+      }
+      const file = path + "/" + inputFile;
+      let yaml;
+      try {
+        yaml = fs.readFileSync(file, { encoding: "UTF-8" });
+      } catch (err) {
+        // the plugin does not have a config file, go to the next file
+        return;
+      }
+      const loaded = Yaml.load(yaml, { json: true });
+      parsed = deepmerge(parsed, loaded);
+    });
+
+    let clientSettings = {};
+    // eslint-disable-next-line no-unused-vars
+    for (const [category, settings] of Object.entries(parsed)) {
+      for (const [setting, details] of Object.entries(settings)) {
+        if (details.client) {
+          clientSettings[setting] = details.default;
+        }
+      }
+    }
+    const contents = `var CLIENT_SITE_SETTINGS_WITH_DEFAULTS  = ${JSON.stringify(
+      clientSettings
+    )}`;
+
+    fs.writeFileSync(`${this.outputPath}/` + "settings_out.js", contents);
+    built = true;
+  }
+}
+
+module.exports = function siteSettingsPlugin(...params) {
+  return new SiteSettingsPlugin(...params);
+};
+
+module.exports.parsePluginClientSettings = function (discourseRoot) {
+  let settings = [discourseRoot + "/config"];
+
+  if (shouldLoadPluginTestJs()) {
+    settings = settings.concat(glob.sync(discourseRoot + "/plugins/*/config"));
```

Comment:
For consistency with other plugin-enumerating stuff, we should only include plugin directories which have a `plugin.rb`, and we should exclude plugins which have directory names starting with a `.`.

Easiest thing is probably to lean on the [`discourse-plugins` addon](https://github.com/discourse/discourse/blob/91b991320c267aad6ba88dde26f9fe5729508689/app/assets/javascripts/discourse-plugins/index.js#L86-L94) like this:

https://github.com/discourse/discourse/blob/91b991320c267aad6ba88dde26f9fe5729508689/app/assets/javascripts/discourse/lib/bootstrap-json/index.js#L384-L399

(we'll need to pass a reference to `app` into this function from `ember-cli-build.js`)